pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -95,31 +106,95 @@ type DNSProxy struct {
+	lookupTargetDNSServer func(w dns.ResponseWriter) (serverIP net.IP, serverPort uint16, addrStr string, err error)
 
 	// this mutex protects variables below this point
 	lock.Mutex
 
-	// allowed tracks all allowed matchNames. These are regexps, even simple
-	// matchNames with no regexp wildcards are compiled. We ensure a unique value
-	// per source policy of a matchName because the RegexpMap handles the
-	// reference counting of unique values but will de-dupe repeats.
+	// allowed tracks all allowed L7 DNS rules by endpointID, destination port,
+	// and L3 Selector. All must match for a query to be allowed.
+	//
+	// matchNames with no regexp wildcards are still compiled, internally.
 	// Note: Simple DNS names, e.g. bar.foo.com, will treat the "." as a literal.
-	// We convert these "." in regexps that only contain dots (and alphanumeric
-	// characters) into a regexp "." literal. This is because the more common use
-	// case will be these simple literals.
-	// To insert a wildcard ".", use .{1} to indicate a single wildcard character.
-	allowed *regexpmap.RegexpMap
+	allowed perEPAllow
 
 	// rejectReply is the OPCode send from the DNS-proxy to the endpoint if the
 	// DNS request is invalid
 	rejectReply int
 }
 
+// perEPAllow maps EndpointIDs to ports + selectors + rules
+type perEPAllow map[uint64]portToSelectorAllow
+
+// portToSelectorAllow maps port numbers to selectors + rules
+type portToSelectorAllow map[uint16]cachedSelectorREEntry
+
+// cachedSelectorREEntry maps port numbers to selectors to rules, mirroring
+// policy.L7DataMap but the DNS rules are compiled into a single regexp
+type cachedSelectorREEntry map[policy.CachedSelector]*regexp.Regexp
+
+// setPortRulesForID sets the matching rules for endpointID and destPort for
+// later lookups. It converts newRules into a unified regexp that can be reused
+// later.
+func (allow perEPAllow) setPortRulesForID(endpointID uint64, destPort uint16, newRules policy.L7DataMap) error {
+	// This is the delete case
+	if len(newRules) == 0 {
+		epPorts := allow[endpointID]
+		delete(epPorts, destPort)
+		if len(epPorts) == 0 {
+			delete(allow, endpointID)
+		}
+		return nil
+	}
+
+	newRE := make(cachedSelectorREEntry)
+	for selector, l7Rules := range newRules {
+		reStrings := make([]string, 0, len(l7Rules.DNS))
+		for _, dnsRule := range l7Rules.DNS {
+			if len(dnsRule.MatchName) > 0 {
+				dnsRuleName := strings.ToLower(dns.Fqdn(dnsRule.MatchName))
+				dnsPatternAsRE := matchpattern.ToRegexp(dnsRuleName)
+				reStrings = append(reStrings, "("+dnsPatternAsRE+")")
+			}
+			if len(dnsRule.MatchPattern) > 0 {
+				dnsPattern := matchpattern.Sanitize(dnsRule.MatchPattern)
+				dnsPatternAsRE := matchpattern.ToRegexp(dnsPattern)
+				reStrings = append(reStrings, "("+dnsPatternAsRE+")")
+			}
+		}
+		re, err := regexp.Compile(strings.Join(reStrings, "|"))
+		if err != nil {
+			return err
+		}
+		newRE[selector] = re
```

Comment:
I had started by keeping the L7DataMap around as well but I realised I don't ever use the DNS rules again, but only the regexes. I'm guessing it's safe to keep the CacheSelector reference since we do it in the redirect object anyway.

Reply:
Yeah, I think so. There was actually a bug in this area recently but Jarno's got a fix out for it: #9617